src/main/java/org/spongepowered/server/launch/VanillaServerMain.java

```diff
@@ -138,15 +144,69 @@ public static void main(String[] args) throws Exception {
     private static boolean downloadMinecraft(Path base, boolean autoDownload) throws IOException, NoSuchAlgorithmException {
         // Make sure the Minecraft server is available, or download it otherwise
         Path path = base.resolve(MINECRAFT_SERVER_LOCAL);
-        if (Files.notExists(path) && (!autoDownload || !downloadVerified(MINECRAFT_SERVER_REMOTE, path))) {
-            return false;
+        if (Files.notExists(path)) {
+            if (!autoDownload) {
+                return false;
+            }
+
+            System.out.println("Downloading the versions manifest...");
+
+            // Download the file with all of the Minecraft versions information
+            JsonObject versions = downloadJsonVerified(MINECRAFT_MANIFEST_REMOTE);
+
+            String versionManifestRemote = null;
+
+            // Find the current version manifest URL
+            for (JsonElement versionInfo : versions.getAsJsonArray("versions")) {
+                JsonObject obj = versionInfo.getAsJsonObject();
+
+                String versionId = obj.get("id").getAsString();
+                if (versionId.equals(MINECRAFT_SERVER_VERSION)) {
+                    versionManifestRemote = obj.get("url").getAsString();
+                    break;
+                }
+            }
+
+            if (versionManifestRemote == null) {
+                throw new RuntimeException("Could not find " + MINECRAFT_SERVER_VERSION + "'s manifest URL");
+            }
+
+            // Find the server URL
+            JsonObject versionManifest = downloadJsonVerified(versionManifestRemote);
+            String serverRemote = versionManifest.get("downloads").getAsJsonObject()
+                    .get("server").getAsJsonObject()
+                    .get("url").getAsString();
+
+            if (!downloadVerified(serverRemote, path)) {
+                return false;
+            }
         }
 
         // Make sure Launchwrapper is available, or download it otherwise
         path = base.resolve(LAUNCHWRAPPER_LOCAL);
         return Files.exists(path) || (autoDownload && downloadVerified(LAUNCHWRAPPER_REMOTE, path));
     }
 
+    private static JsonObject downloadJsonVerified(String remote) throws IOException, NoSuchAlgorithmException {
+        URLConnection con = new URL(remote).openConnection();
+        MessageDigest md5 = MessageDigest.getInstance("MD5");
+        JsonObject json;
+
+        try (DigestInputStream source = new DigestInputStream(con.getInputStream(), md5)) {
+            json = new JsonParser().parse(new InputStreamReader(source, StandardCharsets.UTF_8))
+                    .getAsJsonObject();
+        }
+
+        String expected = getETag(con);
```

Comment:
The ETag isn't guaranteed to be the MD5 of the file (although that appears to be the case in practice).  Not new to your code, but this is still incorrect.  The ETag should be used for caching, but not comparing.

The `downloads` block in the version JSON contains a SHA-1 hash that can be used instead, although I don't think there is any hash for the version JSON itself.

Reply:
You're right, I just copied a portion of code in the #downloadVerified method below.

Reply:
Please just remove the etag stuff entirely. The JSON will just fail to parse if it's invalid.